### ISSUE TITLE
No output from a test which throws an exception in a thread

### ISSUE THREAD

**Issue by paulbutcher** [flix/flix#5459]
I'm creating a new issue for this rather than re-opening old issues, as those issues have lots of discussion which is (I think) no longer relevant.

Once #5458 is merged, #5123 becomes a problem again. The situation isn't quite as bad as it used to be: a test which throws an exception in a thread does now fail, but the stack trace is silently swallowed by the Flix test framework. So the programmer knows that something's failed, but has no clue why.

We discussed different solutions to this at length in #5123. I still, very strongly, dislike the test framework redirecting stderr, and still can't think of a reason why this is desirable behaviour (I know that it's been asserted that "people requested this" but I still haven't seen an example of why it would be desirable).

We need a fix for this. The simplest approach would be to avoid redirecting stderr. But if we can't do that, can we please have a conversation about what the desired behaviour is (and why we chose to redirect stderr in the first place) so that we can design a solution that addresses the use case (whatever it is) without hiding critical information from the programmer.

**Comment by paulbutcher** [comment]
That works for me. But I would still appreciate understanding why this functionality is necessary at all. I'm not doubting that it is, I just don't know what the use case is, which makes it impossible to judge whether a solution is a good or a bad one.

**Comment by magnus-madsen** [comment]
> That works for me. But I would still appreciate understanding why this functionality is necessary at all. I'm not doubting that it is, I just don't know what the use case is, which makes it impossible to judge whether a solution is a good or a bad one.

- In a sea of tests, its important to have a summary of all the relevant information at the end of the output.
- We will (eventually) need to capture this output to integrate it into the VSCode Test API.

**Comment by magnus-madsen** [comment]
Another thought occurred to me.

The problem with `System.exit` is that it does not work well with a test framework or with VSCode (because it takes down the whole process). 

Recalling what Jonathan said, if an `ArrayIndexOutOfBoundsException` occurs in a thread -- assuming our code is correct -- then it gets propagated to the spawning thread and rethrown. This continues until we hit main where (1) the exception is printed (with a more usable stack trace -- I might add) and (2) the program shuts down. Thus this is effectively the same as using `System.exit`. 

Now our shared concern was programmers relying on this to catch exceptions. We wanted to avoid Flix turning into half monadic Result/Validation and half-exception handling. But perhaps we need a change in viewpoint: We could simply __ban__ `catch` from catching any non-checked exception. Thus you can catch `IOException` (good), but not `ArrayIndexOutOfBoundsException` or `MatchError`. Now I am guessing many frameworks come with their own exceptions which are a blend of checked/unchecked with no principles at all, so we probably cannot rely on `Exception/RuntimeException` to decide what you can catch. Instead, I think we have to accept that you can catch everything not black listed. What would be blacklisted? Probably everything `Error`, everything thrown by Flix itself, and select Java exceptions (NPE, AIOOB, etc.).

**Comment by paulbutcher** [comment]
So (independent of the question of what we should be doing with exceptions from child threads) I've been doing some experimentation to see how Flix's test framework behaves. It seems that:

* Output to StdOut is always silently discarded
* Output to StdErr is *sometimes* silently discarded

So this test passes and silently discards the output:

```
@Test
def test48(): Bool \ IO =
    println("this test passes");
    true
```

This test fails, but also silently discards the output:

```
@Test
def test49(): Bool \ IO =
    println("this test fails");
    false
```

This test fails, and displays *both* StdErr *and* the exception stack trace:

```
@Test
def test47(): Bool \ IO =
    System/StdErr.println("report bad thing");
    String.concat("", unsafe_cast null as String) == ""
```

This test fails with "Std Err Output" and then displays StdErr

```
@Test
def test46(): Bool \ IO = 
    System/StdErr.println("this test passes");
    true
```

This test fails, and silently discards StdErr:

```
@Test
def test45(): Bool \ IO = 
    System/StdErr.println("bad thing");
    false
```

But this test fails, then outputs StdErr (and then outputs an unhelpful HoleError stack trace):

```
@Test
def test43(): Bool \ IO = 
    System/StdErr.println("bad thing");
    Assert.eq("this", "that")
```

I can't see any reason why we would ever want to silently discard output? I always want to see all output, always (whether to StdErr or StdOut). What's the benefit in not doing so?

**Comment by magnus-madsen** [comment]
The logic is here:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/tools/Tester.scala#L190

**Comment by paulbutcher** [comment]
Sorry @magnus-madsen, I'm not sure what your comment is intended to convey. That logic does indeed express the behaviour I've observed, but I'm not sure if you're suggesting that it's correct, or incorrect?

Do you agree with:

> I can't see any reason why we would ever want to silently discard output? I always want to see all output, always (whether to StdErr or StdOut). What's the benefit in not doing so?

**Comment by magnus-madsen** [comment]
I'll reply later. I was just identifying the problematic code.

**Comment by mlutze** [comment]
> Another thought occurred to me.
> 
> The problem with `System.exit` is that it does not work well with a test framework or with VSCode (because it takes down the whole process).
> 
> Recalling what Jonathan said, if an `ArrayIndexOutOfBoundsException` occurs in a thread -- assuming our code is correct -- then it gets propagated to the spawning thread and rethrown. This continues until we hit main where (1) the exception is printed (with a more usable stack trace -- I might add) and (2) the program shuts down. Thus this is effectively the same as using `System.exit`.
> 
> Now our shared concern was programmers relying on this to catch exceptions. We wanted to avoid Flix turning into half monadic Result/Validation and half-exception handling. But perhaps we need a change in viewpoint: We could simply **ban** `catch` from catching any non-checked exception. Thus you can catch `IOException` (good), but not `ArrayIndexOutOfBoundsException` or `MatchError`. Now I am guessing many frameworks come with their own exceptions which are a blend of checked/unchecked with no principles at all, so we probably cannot rely on `Exception/RuntimeException` to decide what you can catch. Instead, I think we have to accept that you can catch everything not black listed. What would be blacklisted? Probably everything `Error`, everything thrown by Flix itself, and select Java exceptions (NPE, AIOOB, etc.).

Isn't `Error` vs `Exception` already supposed to be the division between fatal and recoverable?

**Comment by magnus-madsen** [comment]
> Isn't Error vs Exception already supposed to be the division between fatal and recoverable?

I don't think you can meaningfully recover from `IndexOutOfBounds`. It really does represent a situation that was unexpected-- hence not recoverable.

**Comment by magnus-madsen** [comment]
> Sorry @magnus-madsen, I'm not sure what your comment is intended to convey. That logic does indeed express the behaviour I've observed, but I'm not sure if you're suggesting that it's correct, or incorrect?
> 
> Do you agree with:
> 
> > I can't see any reason why we would ever want to silently discard output? I always want to see all output, always (whether to StdErr or StdOut). What's the benefit in not doing so?

I would change it to output stdout+stderr whenever the test fails.

**Comment by magnus-madsen** [comment]
Any case, in the future with AE, tests will not be able to print to stdout/stderr without their effect signature revealing it and requiring a handler (either as part of the mock code or in the test framework). Hence whatever we decide now will likely not be the ultimate solution.

**Comment by mlutze** [comment]
> > Isn't Error vs Exception already supposed to be the division between fatal and recoverable?
> 
> I don't think you can meaningfully recover from `IndexOutOfBounds`. It really does represent a situation that was unexpected-- hence not recoverable.

I disagree. `Exception` being recoverable and `Error` being unrecoverable is part of the API, meaning that Java library authors are "allowed" to throw AIOOB exceptions and expect dependents to catch them.

GitHub code search has [65k results](https://cs.github.com/?scopeName=All+repos&scope=&q=%2Fcatch%5Cs*%5C%28ArrayIndexOutOfBoundsException%2F) for catching AIOOB errors. Of these, many could have just bound-checked beforehand, and many just rethrow, but the remainder I think display a behavior that would be hard to express if we don't allow catching AIOOB.

**Comment by paulbutcher** [comment]
> I would change it to output stdout+stderr whenever the test fails.

Why only when the test fails? Why would I have a test that outputs something and not want to see what it outputs?

**Comment by paulbutcher** [comment]
@mlutze Can I check which of the following you mean:

1. That it is legitimate for Java code to throw AIOOB exceptions, and we should allow Flix programmers to catch and handle them
2. That it is legitimate for well written idiomatic Flix code to throw AIOOB exceptions, and we should allow the Flix programmer to catch and handle them

FWIW, I definitely agree with 1. I can't currently think of any reason why Flix code would legitimately throw an AIOOB exception though.

**Comment by magnus-madsen** [comment]
I think we are running a bit in circles. In the interest of progress. I propose:

- Modify `Tester` to use a proxy for stdout/stderr that immediately prints and captures the output.
- When all tests have completed, print stdout/stderr in the summary, but only for failed tests. (NOTE that the stdout _has already been printed earlier when the test ran_). Reasoning: You probably don't care about random output _in the summary_, but the output is still there, just scroll to the relevant test.

Still contested: Whether to call `System.exit` since it breaks the ability for VSCode interop etc. and/or whether to modify how `catch` works.
